Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,254 @@
+  | succ =>
+    apply Nat.Partrec.succ
+  | left =>
+    apply Nat.Partrec.left
+  | right =>
+    apply Nat.Partrec.right
+  | oracle g =>
+    rw [← h]
+    apply Nat.Partrec.zero
+  | pair _ _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.pair
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  | comp _ _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.comp
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  | prec _ _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.prec
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  | rfind _ _ _ ih =>
+    apply Nat.Partrec.rfind
+    apply ih
+    rw [← h]
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything (f : ℕ →. ℕ) : Nat.Partrec f ↔ ∀ g, RecursiveIn f g :=
+  ⟨(·.recursiveIn), (· _ |>.partrec_of_zero)⟩
+
+/--
+Proof that turing reducibility is reflexive.
+-/
+@[refl]
+theorem RecursiveIn.refl (f : ℕ →. ℕ) : RecursiveIn f f :=
+  RecursiveIn.oracle f
+
+/--
+Instance declaring that `RecursiveIn` is reflexive.
+-/
+instance : IsRefl (ℕ →. ℕ) RecursiveIn :=
+  ⟨RecursiveIn.refl⟩
+
+/--
+Proof that turing reducibility is transitive.
+-/
+theorem RecursiveIn.trans {f g h : ℕ →. ℕ} (hg : RecursiveIn f g) (hh : RecursiveIn g h) :
+    RecursiveIn f h := by
+  induction hg with
+  | zero =>
+    apply RecursiveIn.zero
+  | succ =>
+    apply RecursiveIn.succ
+  | left =>
+    apply RecursiveIn.left
+  | right =>
+    apply RecursiveIn.right
+  | oracle =>
+    exact hh
+  | pair f' h' _ _ _ hf_ih hh_ih =>
+    apply RecursiveIn.pair
+    · apply hf_ih
+      apply hh
+    · apply hh_ih
+      apply hh
+  | comp f' h' _ _ _ hf_ih hh_ih =>
+    apply RecursiveIn.comp
+    · apply hf_ih
+      apply hh
+    · apply hh_ih
+      apply hh
+  | prec f' h' _ _ _ hf_ih hh_ih =>
+    apply RecursiveIn.prec
+    · apply hf_ih
+      apply hh
+    · apply hh_ih
+      apply hh
+  | rfind f' _ _ hf_ih =>
+    apply RecursiveIn.rfind
+    · apply hf_ih
+      apply hh
+
+instance : IsTrans (ℕ →. ℕ) RecursiveIn :=
+  ⟨@RecursiveIn.trans⟩
+
+instance : IsPreorder (ℕ →. ℕ) RecursiveIn where
+  refl := RecursiveIn.refl
+  trans := @RecursiveIn.trans
+
+theorem TuringEquivalent.equivalence : Equivalence TuringEquivalent :=
+  (AntisymmRel.setoid _ _).iseqv
+
+@[refl]
+theorem TuringEquivalent.refl (f : ℕ →. ℕ) : f ≡ᵀ f :=
+  Equivalence.refl equivalence f
+
+@[symm]
+theorem TuringEquivalent.symm {f g : ℕ →. ℕ} (h : f ≡ᵀ g) : g ≡ᵀ f :=
+  Equivalence.symm equivalence h
+
+@[trans]
+theorem TuringEquivalent.trans {f g h : ℕ →. ℕ} (h1 : f ≡ᵀ g) (h2 : g ≡ᵀ h) : f ≡ᵀ h :=
+  Equivalence.trans equivalence h1 h2
+
+/-- The type of Turing degrees, implemented as the
+quotient of partial functions (`PFun`) under Turing equivalence. -/
+abbrev TuringDegree : Type _ :=
+  Antisymmetrization _ RecursiveIn
+
+/--
+Instance declaring that `TuringDegree` is a partially ordered type.
```

Comment:
This isn't strictly accurate - it's giving it a partial order, not declaring that it has one